macrobot/macrobot/handler.go

```diff
@@ -28,19 +33,30 @@ var whiteListedCommands = [...]string{
 
 func NewHandler(stats *base.StatsRegistry, kbc *kbchat.API, debugConfig *base.ChatDebugOutputConfig, db *DB) *Handler {
 	return &Handler{
-		DebugOutput: base.NewDebugOutput("Handler", debugConfig),
-		stats:       stats.SetPrefix("Handler"),
-		kbc:         kbc,
-		db:          db,
+		DebugOutput:  base.NewDebugOutput("Handler", debugConfig),
+		stats:        stats.SetPrefix("Handler"),
+		kbc:          kbc,
+		db:           db,
+		newConvCache: make(map[string]struct{}),
 	}
 }
 
 func (h *Handler) HandleNewConv(conv chat1.ConvSummary) error {
-	// When we're put into a team conv, tell the team about the `create-for-channel` option.
-	if conv.Channel.MembersType == "team" && conv.IsDefaultConv {
+	h.Lock()
+	defer h.Unlock()
+
+	// When we're put into a team conv, tell the team about the
+	// `create-for-channel` option.
+	if _, ok := h.newConvCache[conv.Channel.Name]; !ok && conv.Channel.MembersType == "team" {
 		if err := h.doPrivateAdvertisement(conv.Channel, conv.Id); err != nil {
 			h.Errorf("unable to advertise on new conv: %v", err)
 		}
+		h.newConvCache[conv.Channel.Name] = struct{}{}
+		go func() {
+			// cleanup the cache once we know about all the channels
+			time.Sleep(5 * time.Minute)
+			delete(h.newConvCache, conv.Channel.Name)
```

Comment:
Can this be done without a lock? It might be easier to use a `sync.Map` in this circumstance, I'm not sure. You could probably just get a lock in this goroutine as well if it's needed.

Reply:
lock is for future safety in case we process the requests in separate routines.

Reply:
That makes sense, but couldn't there be a race since no lock is taken for the delete?

Reply:
oh that is an oversight, i'll lock there too!